R/binarise.R

```diff
@@ -0,0 +1,253 @@
+#GB193 multistate 0: they cannot be used attributively, 1: ANM-N; 2: N-ANM; 3: both.
+#GB203 multistate 0: no UQ, 1: UQ-N; 2: N-UQ; 3: both.
+binarise_GBXXX_to_GBXXXa_with_zero <- function(values) {
+    dplyr::case_match(values, "0"~"0", "1" ~ "1", "2" ~ "0", "3" ~ "1", "?" ~ "?",  NA ~ NA)
+}
+
+binarise_GBXXX_to_GBXXXb_with_zero <- function(values) {
+    dplyr::case_match(values, "0"~"0", "1" ~ "0", "2" ~ "1", "3" ~ "1",  "?" ~ "?", NA ~ NA)
+}
+
+
+
+
+gb_recode <- function(ValueTable, oldvariable, newvariable, func) {
+    ValueTable %>% dplyr::filter(Parameter_ID==oldvariable) %>%
+        dplyr::mutate(
+            ID=paste0(newvariable, "-", Language_ID),
+            Parameter_ID=newvariable,
+            Value=func(Value)
+        ) %>%
+        mutate(Code_ID = paste0(Parameter_ID, "-", Value)) %>%
+        rbind(ValueTable)
+}
+
+ValueTable <- gb_recode(ValueTable, 'GB024', 'GB024a', binarise_GBXXX_to_GBXXXa_without_zero)
+
+
+#' Makes multi-state Grambank-features binary in the appropriate manner.
+#'
+#' @param ValueTable data frame of the ValueTable from grambank-cldf (long).
+#' @param drop_multistate If TRUE,the multistate parent features of the binarised features are kept. If FALSE, they are dropped from the resulting data-frame.
+#' @param keep_raw_binary If TRUE and if the value table already contains some binarised features, they are kept. If false, they are overriden and replaced by values derived from the multistate features.
+#' @return Data-frame (long or wide depending on 'wide' argument)
+#' @export
+binarise <- function(ValueTable = NULL, wide = TRUE, drop_multistate = TRUE, keep_raw_binary = TRUE){
+    if (!inherits(ValueTable, "data.frame")) stop("'ValueTable' must be a dataframe.")
+
+    # if there are binary features coded directly by coders, i.e. not derived from multistate features,
+    # we call these "raw binary". users can choose to drop these and only used derived, or use a mix,
+    if (keep_raw_binary == FALSE) {
+        ValueTable <- ValueTable %>%
+            dplyr::anti_join(dplyr::select(binary_parameters_df, Parameter_ID = ID), by = dplyr::join_by(Parameter_ID))
+    }
+
+    # making the long table wide, it's easier for the binarisation. The table can be made long again with
+    # comments etc later.
+    wide_values <- as.grambank.wide(ValueTable)
+
+    #if the binary raw feature isn't there, make a col with just NA values to start with. This will allow us to       blend raw and derived binary data later.
+    for (binary_col in binary_parameters_df$ID) {
+        if(!(binary_col %in% colnames(wide_values))){
+            wide_values[[paste0(binary_col)]] <- rep(NA, nrow(wide_values))
+        }
+    }
+
+    ######### DERIVING binary values from multistate ones
+    wide_values$GB024a <- binarise_gb024_to_gb024a(wide_values$GB024)
+    wide_values$GB024b <- binarise_gb024_to_gb024b(wide_values$GB024)
+
+    # are we deliberately using dplyr::if_else AND base::ifelse?
```

Comment:
@SimonGreenhill yes we were in grambank-analysed because of reasons to do with NAs.
I was told to not spend too much time perfecting that codebase but just make it work, so I made changes to make it work but the overall code wasn't always as elegant as it could be.

Reply:
This comment was in relation to simon asking about the combo of ifelse and if_else. It's no longer relevant, resolving now